codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +530,66 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
+                try {
+                    try {
+                        if (writing) {
+                            throw new IllegalStateException("write is not re-entrant");
+                        }
+                        // Write the portion of the frame.
+                        writing = true;
+                        needFlush |= payload.write(Math.max(0, allowedBytes));
+                    } catch (Exception e) {
+                        // Mark the state as cancelled, we'll clear the pending queue
+                        // via cancel() below.
+                        cancelled = true;
+                    }
+                    if (!cancelled && payload.size() == 0) {
+                        try {
+                            // This frame has been fully written, remove this frame
+                            // and notify the payload. Since we remove this frame
+                            // first, we're guaranteed that its error method will not
+                            // be called when we call cancel.
+                            pendingWriteQueue.remove();
+                            payload.writeComplete();
+                        } catch (Exception e) {
+                            // Mark the state as cancelled, we'll clear the pending queue
+                            // via cancel() below.
+                            cancelled = true;
+                            // Rethrow the exception.
+                            throw new RuntimeException(e);
+                        }
+                    }
+                } finally {
+                    writing = false;
+                    // Make sure we always decrement the flow control windows
+                    // by the bytes written.
+                    writtenBytes = before - payload.size();
+                    decrementFlowControlWindow(writtenBytes);
+
+                    // If a cancellation occurred while writing, call cancel again to
+                    // clear and error all of the pending writes.
+                    if (cancelled) {
+                        cancel();
+                    } else {
+                        // We must not call this in case of error, as cancel() will call it
+                        // for us.
+                        decrementPendingBytes(writtenBytes);
```

Comment:
No, I think we want to do this is all cases.  The call in `onError` decrements by the remaining size of the frame.  So we want to first call `decrementPendingBytes` above to decrement by the amount written and then let `onError` decrement it by the rest.